### PR TITLE
Add contributing guidelines

### LINKED ISSUE #76: add guidelines for contributors

### PR DESCRIPTION
**Pull request by sumeet-bansal**
Closes #76.

### CODE REVIEW

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -1,2 +1,2 @@
-*	@sumeet-bansal
+* @sumeet-bansal
 * @shravanhariharan2
```
Comment (shravanhariharan2) [inline .github/CODEOWNERS:2]: another small nit: these are supposed to be on the same line (see [Github docs](https://docs.github.com/en/free-pro-team@latest/github/creating-cloning-and-archiving-repositories/about-code-owners)). I think that's why after rewrite you weren't getting auto-requested for reviews but I was

**Comment by sumeet-bansal** [comment]
@shravanhariharan2 ready for review, make sure to look at the Markdown files in rich text mode. Anything I forgot to cover or anything confusing in what I did cover?

Gonna let this sit for a couple days and then do a final pass with fresh eyes before merging.

File: .github/CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
+## Contributing
+The membership portal is an open-source project, for members by members, and we welcome any contributions! This project is maintained by ACM UCSD's development team.
```
Comment (shravanhariharan2) [inline .github/CONTRIBUTING.md:2]: nit:
> ... is an open-source project, for members by members, and ...

sounds somewhat clunky. Maybe try:

> The membership portal is an open-source project made for members by members, and we welcome any contributions!"

File: .github/CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+There're a lot of moving parts in this codebase and making simple changes can be daunting as a first-time contributor, so this section breaks down all the areas of the codebase you might need to edit to make your changes.
+
+#### API Specification
+You can find our API specification [here](https://documenter.getpostman.com/view/12949536/TVRedVwQ). It'll tell you about all the endpoints our API exposes and what each does. All the types in our specification, e.g. requests, responses, user profiles, merch orders, are defined in the `types` folder, and if you're adding a route or modifying an existing one, you'll first need to edit these types.
+
+#### API Implementation
+At a high level, our API is implemented in three layers: the endpoints layer, the services layer, and the data layer. The endpoints layer calls the services layer and the servies layer calls the data layer.
+
+##### Endpoints
+If you make a request to some endpoint, that request will be received in our endpoints layer, which handles anything that has to do with our routing logic, with requests and responses. This includes
```
Comment (shravanhariharan2) [inline .github/CONTRIBUTING.md:23]: nit: "Endpoints" in rich-text mode looks smaller than the text below it (even if it's bolded). Maybe up all the header fonts by 1? Like you can just remove 1 # from every header

File: .github/CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+Once your functionality is implemented, open up a PR and a maintainer will review it as soon as possible. Anecdotally, most of our bigger PRs go through several rounds of review before merging, so if you're implementing a feature for an API client, be sure to start early, start often to meet your team's deadlines. If you're stuck or confused at any point, please don't hesitate to ask one of our maintainers for help.
+
+### Project Structure
+There're a lot of moving parts in this codebase and making simple changes can be daunting as a first-time contributor, so this section breaks down all the areas of the codebase you might need to edit to make your changes.
+
+#### API Specification
+You can find our API specification [here](https://documenter.getpostman.com/view/12949536/TVRedVwQ). It'll tell you about all the endpoints our API exposes and what each does. All the types in our specification, e.g. requests, responses, user profiles, merch orders, are defined in the `types` folder, and if you're adding a route or modifying an existing one, you'll first need to edit these types.
```
Comment (shravanhariharan2) [inline .github/CONTRIBUTING.md:17]: I think you should bold the sentence "All the types in our specification... defined in the `types` folder", since I think it's super important to highlight that `types` should be the starting point for any large feature update to a newcomer

File: .github/CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
++ checking permissions,
++ calling the services layer with any relevant request data,
++ and sending back the response.
+
+This layer lives inside our `api` folder. In the unlikely event you'll need to write middleware, that'll go in the `middleware` folder. The `controllers` folder is where the bulk of our routing logic lives; within each controller, you'll see the API routes corresponding to that controller, e.g. `UserController` holds the routes under `/api/user`. Most of these routes require some sort of request body&mdash;the same request bodies you specified in the `types` folder&mdash;, and these request bodies need to be validated; to do so, create classes in the `validators` folder implementing the interfaces from the `types` folder. That leaves the actual endpoint code in the controllers, which should take any relevant data from the requests and call the services layer. Occasionally, requests have to be further validated in the controller, such as to ensure there're no duplicate or out-of-order elements in an array.
```
Comment (shravanhariharan2) [inline .github/CONTRIBUTING.md:30]: nit: split this paragraph into two at "Most of these routes require...", that way its more readable and we're highlighting validation even more

File: .github/CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
++ [Express](https://github.com/expressjs/express) is a web framework we use
++ [routing-controllers](https://github.com/typestack/routing-controllers) is a library we use on top of Express
++ [class-validator](https://github.com/typestack/class-validator) is a library we use for decorator-based validation
+
+##### Services
+The services layer is made up of all the actual "business logic" of the portal, the rules of the system (e.g. "a user can only attend an event once"). Each service is some grouping of functionality; for example, the `EventService` contains all code for creating events, reading events, etc. All our services can be found in the `services` folder.
```
Comment (shravanhariharan2) [inline .github/CONTRIBUTING.md:37]: nit: 
> ... "business logic" of the portal, the rules of the system ..."

might be a fragment? I think replacing comma with "and" makes more sense

File: .github/CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
...schema requires not only direct edits to the data models (found in the `models` folder) but also schema migrations (found in the `migrations` folder). Schema migrations are procedures for changing the schema from an older version (such as the schema currently running in production) to your modified version. To write a schema migration, take a look at the [relevant TypeORM documentation](https://github.com/typeorm/typeorm/blob/master/docs/migrations.md) and [existing migrations](https://github.com/acmucsd/membership-portal/tree/master/migrations). Everything's already configured via [`ormconfig.ts`](https://github.com/acmucsd/membership-portal/blob/master/ormconfig.ts) so running `npm run db:migrate` will execute your migration(s). Finally, we have our data repositories, which contain predefined data queries that abstract away database interactions such as complex SQL queries and can be easily called by services, in the `repositories` folder. All data in use by the application should be immutable, and the services should not interact with the database except through repository methods. This means preferring `userRepository.update(user, changes)` to `user.field = value; user.save()`.
```
Comment (shravanhariharan2) [inline .github/CONTRIBUTING.md:46]: Maybe bold the first sentence? I think generally there should be more bolding since I'm not sure how many people would read every single paragraph. Important things like validation, migrations, transactions, and others I think are worth bolding for importance

File: .github/CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
++ [TypeORM](https://github.com/typeorm/typeorm/) is a library we use to manage our data models
+
+#### Testing
+"Software engineering is the integral of coding over time"; to build a reliable system over several years, we rely on tests to verify that we've implemented changes correctly without breaking anything unexpected in the process, that we throw errors when appropriate, and that we've handled forgettable edge cases. This is currently a bit of a work in progress, but it's something to keep in mind as you're developing&mdash;the use cases to be tested&mdash;and once we've established some examples and work towards testing the entire system, we'll expect you to write tests for your own PRs.
```
Comment (shravanhariharan2) [inline .github/CONTRIBUTING.md:51]: I think we should add a note on documentation as well. We should mention that routes should be documented and that contributors should message us for the Postman credentials

File: README.md [hunk 9]
```diff
@@ -1,5 +1,5 @@
 ## membership-portal-api &nbsp; [![CircleCI](https://circleci.com/gh/acmucsd/membership-portal/tree/master.svg?style=svg)](https://circleci.com/gh/acmucsd/membership-portal/tree/master)
-REST API for the UC San Diego ACM chapter's membership portal.
+REST API for the UC San Diego ACM chapter's membership portal. This is an open-source project, for members by members, and we welcome any contributions! If you're interested in contributing or using the API for your own project, check out our guide [here](https://github.com/acmucsd/membership-portal/blob/master/.github/CONTRIBUTING.md).
```
Comment (shravanhariharan2) [inline README.md:2]: nit: same thing w/ for members by members

File: .github/CONTRIBUTING.md (hunk 4 above)
Comment (sumeet-bansal) [reply .github/CONTRIBUTING.md:17]: The project structure section's in chronological order for changes to make. I mostly bolded sentences in the Getting Started section since people might skim over that, but if you're a newcomer you'll probably read the project structure section in its entirety.

File: .github/CONTRIBUTING.md (hunk 7 above)
Comment (sumeet-bansal) [reply .github/CONTRIBUTING.md:46]: There's no bolding in the project section because first-time contributors should read all of it.

**Comment by sumeet-bansal** [comment]
Updated, but gonna solicit some feedback from Sage and Dimensions too.

**Review by shravanhariharan2 — APPROVED** [review]
Are we still waiting for anyone else's eyes on this? @michl1001 feel free to suggest any edits for additional clarity

**Comment by michl1001** [comment]
In the readme, it looks like the instructions for Macs, Linux, and Windows are all on one line, it might help to have them has separated bullets

Outside this, looks fine to me!
